orttraining/orttraining/python/training/orttrainer.py

```diff
@@ -843,3 +854,306 @@ def _update_onnx_model_initializers(self, state_tensors):
+
+        # add partition information in case of a distributed run
+        if self.options.distributed.deepspeed_zero_optimization.stage > 0:
+            partition_info = 'partition_info'
+            states[partition_info] = self._training_session.get_partition_info_map()
+
+        return states
+
+    def _load_model_states(self, training_session_state_dict, state_dict, strict):
+        """Load the model states onto the training session state dictionary"""
+
+        model = 'model'
+
+        # collect all initializer names from the current onnx graph
+        initializer_names = {node.name for node in self._onnx_model.graph.initializer}
+
+        # loaded_initializers dict will be loaded with all the model states from the state dictionary
+        # that are found in the initializer_names dictionary
+        loaded_initializers = {}
+
+        # create an entry for the model in the training session state dictionary
+        if model not in training_session_state_dict:
+            training_session_state_dict[model] = {}
+
+        # copy over model states from the input state dict onto the training session state dict
+        # and copy over model states from the input state dict onto the onnx model
+        for precision, precision_states in state_dict[model].items():
+            if precision not in training_session_state_dict[model]:
+                training_session_state_dict[model][precision] = {}
+
+            for state_key, state_value in precision_states.items():
+                state_value_np = torch.tensor(state_value).numpy()
+                training_session_state_dict[model][precision][state_key] = state_value_np
+                if state_key in initializer_names:
+                    loaded_initializers[state_key] = state_value_np
+                elif strict:
+                    raise RuntimeError("Unexpected key: {} in state_dict[model][{}]".format(state_key, precision))
+
+        # update onnx model from loaded initializers
+        self._update_onnx_model_initializers(loaded_initializers)
+
+    def _load_optimizer_states(self, training_session_state_dict, state_dict):
+        """Load the optimizer states onto the training session state dictionary"""
+
+        optimizer = 'optimizer'
+
+        # create an entry for the optimizer in the training session state dictionary
+        if optimizer not in training_session_state_dict:
+            training_session_state_dict[optimizer] = {}
+
+        # copy over optimizer states from the input state dict onto the training session state dict
+        for model_state_key, optimizer_dict in state_dict[optimizer].items():
+            if model_state_key not in training_session_state_dict[optimizer]:
+                training_session_state_dict[optimizer][model_state_key] = {}
+            for optimizer_state_key, optimizer_state_value in optimizer_dict.items():
+                optimizer_state_value_np = torch.tensor(optimizer_state_value).numpy()
```

Comment:
is the only reason for this copy the need to maintain the state in numpy format?

Reply:
yes, the value can be either a numpy array or a tensor. So, need a way to convert to a numpy array before passing it to the backend.

Reply:
again, may be we can check if there is actually a need to convert from torch tensor to numpy format. Right now, we have been able to pass a torch state_dict to ORT backend successfully so this might be unnecessary.

Reply:
I checked. It appears like we need to use numpy arrays to populate the onnx model on the frontend. So, the conversion is required here.
Alternatively (as I have done now), the returned state dictionary will be a numpy array if ```pytorch_format=False``` and a torch tensor if ```pytorch_format=True```